akka-http/src/main/scala/akka/http/scaladsl/coding/NegotiableDecoder.scala

```diff
@@ -0,0 +1,17 @@
+package akka.http.scaladsl.coding
+
+import akka.http.scaladsl.model.HttpResponse
+import akka.http.scaladsl.model.headers.{HttpEncoding, HttpEncodings}
+import akka.stream.scaladsl.Flow
+
+object NegotiableDecoder {
+  private def findCodec(encoding: HttpEncoding) = encoding match {
+    case HttpEncodings.gzip | HttpEncodings.`x-zip` => Gzip
+    case HttpEncodings.deflate => Deflate
+    case _ => NoCoding
+  }
+
+  def decode(resp: HttpResponse) = findCodec(resp.encoding).decode(resp)
+
+  def flow() = Flow[HttpResponse].map(decode)
+}
```

Comment:
In general I guess it'll be something around that style, yeah.
We should also allow picking a specific one in the same API.
Maybe we could make it more fluent, as in `via.(Coding.decode: Flow)`?